### DISCUSSION TITLE
Support for C++20 <ranges> library

### DISCUSSION THREAD

**Discussion by evilenzo** [nlohmann/json#3408] [Ideas]
I met a problem that I need to iterate through reversed json array. What I'd like to use in C++20 is
```C++
for (auto i: my_json["some_array"] | std::views::reverse) {
    // do something
} 
```
Is it such a problem to implement this kind of feature?

**Comment by nlohmann** [comment]
In order to asses your issue, we need the following information:

- What is the issue you have?

- Please describe the steps to reproduce the issue. Can you provide a small but working code example?

- What is the expected behavior?

- And what is the actual behavior instead?

- Which compiler and operating system are you using? Is it a [supported compiler](https://github.com/nlohmann/json#supported-compilers)?

- Did you use a released version of the library or the version from the `develop` branch?

- If you experience a compilation error: can you [compile and run the unit tests](https://github.com/nlohmann/json#execute-unit-tests)?

**Comment by gregmarr** [comment]
> Is it such a problem to implement this kind of feature?

No, this works just fine, with `std::views::reverse` instead of `std::views::reversed`, at least with GCC trunk.

[Compiler Explorer](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,selection:(endColumn:79,endLineNumber:10,positionColumn:79,positionLineNumber:10,selectionStartColumn:79,selectionStartLineNumber:10,startColumn:79,startLineNumber:10),source:'%23include+%3Cmemory%3E%0A%23include+%3Ciostream%3E%0A%23include+%3Calgorithm%3E%0A%23include+%3Cvector%3E%0A%23include+%3Cranges%3E%0A%23include+%3Cnlohmann/json.hpp%3E%0A%0Aint+main()%0A%7B%0A++++auto+my_json+%3D+nlohmann::json::parse(%22%7B+%5C%22some_array%5C%22:+%5B1,+2,+3,+4,+5%5D%7D%22)%3B%0A%0A++++std::cout+%3C%3C+my_json.dump()+%3C%3C+%22%5Cn%22%3B%0A%0A++++for+(auto+i:+my_json%5B%22some_array%22%5D+%7C+std::views::reverse)+%7B%0A++++++++std::cout+%3C%3C+i+%3C%3C+%22%5Cn%22%3B%0A++++%7D+%0A%0A++++return+0%3B%0A%7D%0A'),l:'5',n:'0',o:'C%2B%2B+source+%231',t:'0')),k:48.07164860368383,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((g:!((h:compiler,i:(compiler:gsnapshot,filters:(b:'0',binary:'0',commentOnly:'0',demangle:'0',directives:'0',execute:'0',intel:'0',libraryCode:'0',trim:'1'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,libs:!((name:nlohmann_json,ver:trunk)),options:'-std%3Dc%2B%2B20',selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1,tree:'1'),l:'5',n:'0',o:'x86-64+gcc+(trunk)+(C%2B%2B,+Editor+%231,+Compiler+%231)',t:'0')),header:(),k:38.79508502582705,l:'4',m:25.533980582524272,n:'0',o:'',s:0,t:'0'),(g:!((h:output,i:(compiler:1,editor:1,fontScale:14,fontUsePx:'0',tree:'1',wrap:'1'),l:'5',n:'0',o:'Output+of+x86-64+gcc+(trunk)+(Compiler+%231)',t:'0')),header:(),l:'4',m:74.46601941747572,n:'0',o:'',s:0,t:'0')),k:51.92835139631617,l:'3',n:'0',o:'',t:'0')),l:'2',n:'0',o:'',t:'0')),version:4)

**Reply by falbrechtskirchinger** [reply]
To my knowledge no release version works because of the iterator issue I fixed last month. I had trouble using `std::ranges` algorithms before.

Edit: Just checked with 3.10.5 and develop. The former fails to compile with an undecipherable error message.

**Reply by gregmarr** [reply]
Ah, I was using trunk.

**Reply by evilenzo** [reply]
Of course, I was talking about `std::views::reverse`, my mistake

**Comment by falbrechtskirchinger** [comment]
Try using an unreleased version after commit 700b95f4473cac1d900b807541daad0ca3d98f52. That should work.

Edit: @nlohmann You might want to mention that specifically in the 3.10.6 release notes because almost no one will read "Make iterator operator++/--(int) equality-preserving" and think that that means support for `std::ranges`. Also I should contribute some unit tests to be sure most things really work.

**Reply by nlohmann** [reply]
There is also #3130 - it would be great to have this fixed together.

**Reply by falbrechtskirchinger** [reply]
> There is also #3130 - it would be great to have this fixed together.

I imagine it is but I haven't checked yet. Might do so later.

**Reply by gregmarr** [reply]
> I imagine it is but I haven't checked yet. Might do so later.

Unfortunately, still broken.

>source>:13:24: error: no match for 'operator|' (operand types are 'nlohmann::basic_json<>::iteration_proxy<nlohmann::detail::iter_impl<nlohmann::basic_json<> > >' {aka 'nlohmann::detail::iteration_proxy<nlohmann::detail::iter_impl<nlohmann::basic_json<> > >'} and 'std::ranges::views::__adaptor::_Partial<std::ranges::views::_Transform, main()::<lambda(auto:23&&)> >')
   13 |     auto t = j.items() | std::ranges::views::transform([](auto&& _v) { return _v.key(); })
      |              ~~~~~~~~~ ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                     |                                 |
      |                     |                                 std::ranges::views::__adaptor::_Partial<std::ranges::views::_Transform, main()::<lambda(auto:23&&)> >
      |                     nlohmann::basic_json<>::iteration_proxy<nlohmann::detail::iter_impl<nlohmann::basic_json<> > > {aka nlohmann::detail::iteration_proxy<nlohmann::detail::iter_impl<nlohmann::basic_json<> > >}

**Reply by falbrechtskirchinger** [reply]
Yeah, I forgot that there are other outstanding issues in the iterator implementation. Too bad.

**Reply by falbrechtskirchinger** [reply]
> > I imagine it is but I haven't checked yet. Might do so later.
> 
> Unfortunately, still broken.
> 
> > source>:13:24: error: no match for 'operator|' (operand types are 'nlohmann::basic_json<>::iteration_proxy<nlohmann::detail::iter_impl<nlohmann::basic_json<> > >' {aka 'nlohmann::detail::iteration_proxy<nlohmann::detail::iter_impl<nlohmann::basic_json<> > >'} and 'std::ranges::views::__adaptor::_Partial<std::ranges::views::_Transform, main()::<lambda(auto:23&&)> >')
> > 13 |     auto t = j.items() | std::ranges::views::transform([](auto&& _v) { return _v.key(); })
> > |              ~~~~~~~~~ ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> > |                     |                                 |
> > |                     |                                 std::ranges::views::__adaptor::_Partial<std::ranges::views::_Transform, main()::<lambda(auto:23&&)> >
> > |                     nlohmann::basic_json<>::iteration_proxy<nlohmann::detail::iter_impl<nlohmann::basic_json<> > > {aka nlohmann::detail::iteration_proxy<nlohmann::detail::iter_impl<nlohmann::basic_json<> > >}

Good news! Thanks to @gregmarr's prior work [here](https://github.com/nlohmann/json/issues/3130#issuecomment-967460905) and a few changes and additions of my own, I'm able to successfully compile and run this line of code now. PR soon.

I only wish I had seen @davidstone's [comment](https://github.com/nlohmann/json/issues/3130#issuecomment-1079283128) in that thread before now. It took me forever to figure out why `std::indirectly_readable` isn't satisfied.
